Enhance error message for Future.setErrorIfNeeded.

pritamdamania87 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#56631 Enhance error message for Future.setErrorIfNeeded.**

`setErrorIfNeeded` did not mention whether the future was already
completed or there was some other exception. This particular change ensures
that we also print out the original exception as part of the error message.

This would help in debugging issues where this codepath is triggered.

Differential Revision: [D27919974](https://our.internmc.facebook.com/intern/diff/D27919974/)

pritamdamania committed:
Enhance error message for Future.setErrorIfNeeded.

`setErrorIfNeeded` did not mention whether the future was already
completed or there was some other exception. This particular change ensures
that we also print out the original exception as part of the error message.

This would help in debugging issues where this codepath is triggered.

Differential Revision: [D27919974](https://our.internmc.facebook.com/intern/diff/D27919974/)

[ghstack-poisoned]

rohan-varma commented on aten/src/ATen/core/ivalue_inl.h:
```diff
@@ -412,10 +413,16 @@ struct C10_EXPORT ivalue::Future : c10::intrusive_ptr_target {
+        msg += fmt::format(
+            ", \nOriginal exception:\n {}",
+            tryRetrieveErrorMessageInternal(eptr_));
+      }
```
nit: If it was completed without an error, should we log that as well (and possibly the IValue it was completed with)? We should usually have an error so it might be good if that happens to add it as a `LOG(WARNING)`.

pritamdamania committed:
Update on "Enhance error message for Future.setErrorIfNeeded."


`setErrorIfNeeded` did not mention whether the future was already
completed or there was some other exception. This particular change ensures
that we also print out the original exception as part of the error message.

This would help in debugging issues where this codepath is triggered.

Differential Revision: [D27919974](https://our.internmc.facebook.com/intern/diff/D27919974/)

[ghstack-poisoned]

pritamdamania committed:
Update on "Enhance error message for Future.setErrorIfNeeded."


`setErrorIfNeeded` did not mention whether the future was already
completed or there was some other exception. This particular change ensures
that we also print out the original exception as part of the error message.

This would help in debugging issues where this codepath is triggered.

Differential Revision: [D27919974](https://our.internmc.facebook.com/intern/diff/D27919974/)

[ghstack-poisoned]
